Feature/add get_anime_search

dark1zinn opened the pull request:
## Adds feature for `get_anime_search` - Resolves https://github.com/Sidharth-Singh10/jikan-rs/issues/9#issue-3049368071

For a better understanding please refer to issue #9 

#### Changes made:
- Added `get_anime_search` in JikanClient impl at [src/anime.rs:183](https://github.com/dark1zinn/jikan-rs/blob/6a7472835755739157a7c294cbd0586913525572/src/anime.rs#L183)
- Added struct `SearchParams` and enum `AnimeType` at [src/anime.rs:117](https://github.com/dark1zinn/jikan-rs/blob/6a7472835755739157a7c294cbd0586913525572/src/anime.rs#L117) and [src/anime.rs:104](https://github.com/dark1zinn/jikan-rs/blob/6a7472835755739157a7c294cbd0586913525572/src/anime.rs#L104)
- Added non pub utility fn `format_search_query` for formatting the string passed in the param `q` at `get_anime_search` to get rid of special characters and replacing spaces at [src/anime.rs:164](https://github.com/dark1zinn/jikan-rs/blob/6a7472835755739157a7c294cbd0586913525572/src/anime.rs#L164)
- Added test for `get_anime_search` at [tests/anime.rs:17](https://github.com/dark1zinn/jikan-rs/blob/6a7472835755739157a7c294cbd0586913525572/tests/anime.rs#L17)

Sidharth-Singh10 commented:
Hi @dark1zinn, thanks for opening the issue and submitting the corresponding PR!

Could you please address the warning generated by clippy as well?

dark1zinn commented:
Sure! I will look on that as soon as possible!

dark1zinn commented:
All done @Sidharth-Singh10!
Just removed the impl for the struct `SearchParams` and added `Default` to its `#[Derive()]`.
The test runned smothly with no errors!

Thunder-Blaze commented:
I am reviewing 👍🏻

Thunder-Blaze commented:
@dark1zinn Great Work, It's working as expecting and the code quality is great

Things to mention
 - rating, status, order_by, sort could also have their own enum like type
 - Tests are a bit incomplete, in my opinion you should've created a test for each query and one with all queries (no queries is not possible since q is required)
 
@Sidharth-Singh10 what's your final take ?

Sidharth-Singh10 commented:
Thanks for the review @Thunder-Blaze
@dark1zinn Could you please create separate enums for **rating**, **status**, **order_by**, and **sort** to maintain consistency with how type is handled?
Also, for the tests, please add one that includes all query parameters combined to ensure full coverage.

Sidharth-Singh10 commented:
Thanks for the review @Thunder-Blaze
@dark1zinn Could you please create separate enums for **rating**, **status**, **order_by**, and **sort** to maintain consistency with how type is handled?
Also, for the tests, please add one that includes all query parameters combined to ensure full coverage.

dark1zinn commented:
@Sidharth-Singh10 I will be doing this right now!
Again, im sorry for my mistake! 🙏

Thunder-Blaze commented:
@dark1zinn no worries
Also there are some of those Enums present in other files.
You can make use of them though they were not implemented as good as you did inside the functions.
You can copy or import them for now until we create a separate file for commonly used Enums

Sidharth-Singh10 commented:
@dark1zinn No worries at all — mistakes are a natural part of the learning process! 🙌
The fact that you're open to feedback and ready to act on it is what really matters.

Thunder-Blaze commented:
> @dark1zinn No worries at all — mistakes are a natural part of the learning process! 🙌 The fact that you're open to feedback and ready to act on it is what really matters.

Yeah, I also learned many new things in rust thanks to your contribution @dark1zinn
I didn't expect it to infer string by enum without specifying but it did, now I realize my previous PRs had useless definations, I'll also improve them now based on your code 🚀

dark1zinn commented:
When running `cargo clippy`, it complains about "ambiguous name" for the enums, pointing to the file [src/people.rs](https://github.com/dark1zinn/jikan-rs/blob/feature/add-get-anime-search/src/people.rs) at lines `104`, `120-122` and `128` due importing with an `*` at top lines.
Should i change it?

Thunder-Blaze commented:
> @Thunder-Blaze thanks! Honestly, i did noticed my fault on the lack of tests for each query, I intentionally skipped doing these because of my own egoism and hurry to have this feature on the crate, im very very sorry for this mistake. Also, the mention about the enums for `rating`, `status`, `order_by` and `sort` is a very good point wich I didn’t think of due my hurry, again i feel embarrassed for this. Thanks a lot for the review, as a newbie in the Rust programming language i feel very excited to receive such feedbacks!



> When running `cargo clippy`, it complains about "ambiguous name" for the enums, pointing to the file [src/people.rs](https://github.com/dark1zinn/jikan-rs/blob/feature/add-get-anime-search/src/people.rs) at lines `104`, `120-122` and `128` due importing with an `*` at top lines. Should i change it?

prefer importing only the ones you need like from recommendations you want to import sort, orderby
`recommendations::{Sort, OrderBy}` rather than `recommendations::*`
If it's error then please do the changes, if it's warning then ig you can let it be

Thunder-Blaze commented:
Oh I see, you redefined them in anime.rs
It won't cause any error then ig

dark1zinn commented:
> Oh I see, you redefined them in anime.rs It won't cause any error then ig

I actually went to `src/people.rs` and explicity imported the ones from `character`
Right now im running the tests, since it returns an error when theres no anime called "Dragon" that is airing and started last year, i need to find a query that actually return somenthing lol.

Thunder-Blaze commented:
> > Oh I see, you redefined them in anime.rs It won't cause any error then ig
> 
> I actually went to `src/people.rs` and explicitly imported the ones from `character`
> Right now im running the tests, since it returns an error when theres no anime called "Dragon" that is airing and started last year, i need to find a query that actually return somenthing lol.

Search for To Be Hero X 😂 
Or maybe AOT or Death Note (for these change from airing to complete)

dark1zinn commented:
True, i'll put the query to match exactly with an anime, that is a genius ideia!

dark1zinn commented:
Unfortunately i couldn't resolve it yet and i have to go out now, so while im out if you find any way on making a query that works properly please leave-me an comment, ill be out for a few hours.

dark1zinn commented:
@Sidharth-Singh10 @Thunder-Blaze 
Alright, should be all good now, please review the PR, anything else just let me know.
I may also add an error handling when the param for `start_date` and `end_date` don't follow the `YYYY-MM-DD` format later...

Thunder-Blaze commented:
Oh I see, you redefined them in anime.rs
It won't cause any error then ig

> I'm back! Plus i found out that `min_score`, `max_score`, `start_date`, `end_date` and `letter` panics when used, i will look on a fix for that right now

when checking for min and max score, set the score to None and then try
I think score and max_score/min_score aren't supposed to be used together

Thunder-Blaze commented:
Umm well on checking properly ig your test is wrong
![image](https://github.com/user-attachments/assets/91d7778a-c60f-4fda-b55f-34c93d16655a)
Use println!() after the result and check that you get at least 1 anime (You can fill the data acc to 1 single anime but there should be atleast 1 anime in the result and after that it should pass the test)

dark1zinn commented:
Now that i think of it, it would be kinda bad if i wanted to search for animes based only on genres and ratings instead of explicitly using an search term/string

dark1zinn commented:
@Thunder-Blaze, to be fair, in the test i was considerating only the fact that the request should not fail with a bad request or somenthing like that, and since through the existing code the assertion for `result.is_ok()` was used to define if the test was a succes, i understood that if the query finds no animes but completes succesfully, the check in the query to verify if any anime was actually found would be done by the user/dev.
Once the result depends on the provided params to actually find somenthing, that would be quite common to happen, or you meant that i should return an error if the data array is empty?

Thunder-Blaze commented:
> @Thunder-Blaze, to be fair, in the test i was considerating only the fact that the request should not fail with a bad request or somenthing like that, and since through the existing code the assertion for `result.is_ok()` was used to define if the test was a succes, i understood that if the query finds no animes but completes succesfully, the check in the query to verify if any anime was actually found would be done by the user/dev. Once the result depends on the provided params to actually find somenthing, that would be quite common to happen, or you meant that i should return an error if the data array is empty?

No I meant for you to tailor the test such that at least 1-2 anime is printed
Because when there is some result returning then only can we check the thing is right
> If you don't get what I mean then run a query which gives no result with Sort::Desc and then with Sort::Asc, now how will you identify if they were working or not when the list is empty (there is nothing to sort so it will always return true)
 
also for some reason ordering by MalId is giving error

Thunder-Blaze commented:
> Now that i think of it, it would be kinda bad if i wanted to search for animes based only on genres and ratings instead of explicitly using an search term/string

Sometimes you can't remember the name but can remember some other details and since the original Jikan API has that feature, adding in ours also won't take much time, just put the q in search params and remove from the function and remove the required condition

dark1zinn commented:
@Thunder-Blaze don't you think that explicitly specifying the enums string on a match case would be better? just so we can avoid these issues....

dark1zinn commented:
![image](https://github.com/user-attachments/assets/a47afcf7-c91b-4504-9b92-5549974107ce)
@Thunder-Blaze, apparently, my concern early was true.

Thunder-Blaze commented:
> ![image](https://github.com/user-attachments/assets/a47afcf7-c91b-4504-9b92-5549974107ce)
> @Thunder-Blaze, apparently, my concern early was true.
> Note that for the `sort` param this does not affect it since it has no words sepparated by an underline `_`, but for a more concise structure i will do for it as well

Ahh so it was just taking the enum as string lol, here I thought it was auto determining which is weird on thinking normally

Thunder-Blaze commented:
> @Thunder-Blaze don't you think that explicitly specifying the enums string on a match case would be better? just so we can avoid these issues....
> 

Yeah it seems like we will have to do this way for now atleast

dark1zinn commented:
![image](https://github.com/user-attachments/assets/e395ce29-e259-425f-abf3-4a9b15783597)
Test is now returning a more acceptable response for test verification

dark1zinn commented:
> > @Thunder-Blaze don't you think that explicitly specifying the enums string on a match case would be better? just so we can avoid these issues....
> 
> Yeah it seems like we will have to do this way for now atleast

I did looked for an alternative instead of using the match case, but didnt found much...

Thunder-Blaze commented:
> ![image](https://github.com/user-attachments/assets/e395ce29-e259-425f-abf3-4a9b15783597)
> Test is now returning a more acceptable response for test verification

Nicee, great work bro 👍🏻

dark1zinn commented:
@Sidharth-Singh10 could you please check the PR for it to be merged?

Sidharth-Singh10 commented:
@dark1zinn Thanks for working on the issue.

dark1zinn commented:
@Sidharth-Singh10 thanks for accepting my PR, i did learn a lot working on this feature.
Can you tell me when this feature will be available in crates.io?

Sidharth-Singh10 commented:
@dark1zinn probably by 12th May night.
If you have some urgent need, say so, I will try to update the crate ASAP

dark1zinn commented:
> @dark1zinn probably by 12th May night. If you have some urgent need, say so, I will try to update the crate ASAP

Thanks for the preucupation @Sidharth-Singh10, i'm in need of this feature for my project wich is a desktop app, if you allow me, i would like to drop the link for the discord here alongside with some explanation on the purpose of the app.
I´ve already created its organization, [Anime Pack](https://github.com/anime-pack), and if possible for anyone interested in contributing it would help a lot! ❤️

Sidharth-Singh10 commented:
@dark1zinn  Of course, Feel free to drop the Discord link and share more about your app and its purpose.
@Thunder-Blaze currently has a refactor PR open. Once that gets merged and the corresponding update is published to crates.io, I’ll make sure to ping you. I would  recommend waiting until the refactor is complete—shouldn’t take more than 2–3 days at most

dark1zinn commented:
## Anime Pack
An MyAnimeList client on your desktop, for all your animes!

This desktop app comes with an goal not only to let you search info about animes/mangás, and watch/read like the others, our plan is way beyond this, the purpose of building an actual community, with the finnest integration possible with Discord, so all your friends will not only be able to see what youre watching/reading, but to join you in this activity to participate together on laughs and suspenses, not only sharing the moment, but making good memories with friends, not only this, but a lot more!

`
"Please do note that Anime Pack is still in early development at this moment (11/05/2025), and thus despite the will in fully developing this app, to this moment it has only me as a developer and no sponsors/contributors, any kind of help is welcome, we'll be happy just by you joining our community in Discord!"
`
- If you do want to somehow help please open a ticket in the discord server!

For anyone that comes this far (on this merged PR, lol), please feel free to join the [discord server](https://discord.gg/Nv8UXpB36y) of my desktop app, [Anime Pack](https://github.com/anime-pack) (Currently under development of major initial features), to be part of our community, and if you wish to become an active contributor, open up a ticket regarding your will to become an contributor.

### Links:
- Discord: https://discord.gg/Nv8UXpB36y
- Anime Pack organization: https://github.com/anime-pack

> And thanks to these amazing devs for maintaining this wonderful crate!
> @Sidharth-Singh10 
> @Thunder-Blaze 

Made with ❤️ by @dark1zinn.

Sidharth-Singh10 commented:
Great initiative @dark1zinn 
Probably the repos under your organization are private and are not visible to public (us). You may want to check that

dark1zinn commented:
> Great initiative @dark1zinn Probably the repos under your organization are private and are not visible to public (us). You may want to check that

Sure! On the start it wast just quite an test so i was doing everything in private, yesterday i just transferred the repo to the organization, will be opening it right now

Thunder-Blaze commented:
@dark1zinn This is great :D

I was myself also planning to create a desktop app using tauri but I've no experience rn and this is gonna help me learn a lot.

Also I'll try to complete all the remaining endpoints (very less or almost none should be remaining rn) in my open pr in jikan-rs so that you can use them, already completed get_manga_search

Sidharth-Singh10 commented:
Hey @dark1zinn, We have updated the [crate](https://crates.io/crates/jikan_moe). Please let us know if you face any issue
